src/acquisition/covidcast/csv_importer.py

```diff
@@ -0,0 +1,242 @@
+
+      # # extract and validate geographic resolution
+      geo_type = match.group(3).lower()
+      if geo_type not in CsvImporter.GEOGRAPHIC_RESOLUTIONS:
+        print(' invalid geo_type')
+        yield (path, None)
+        continue
+
+      # extract additional values, lowercased for consistency
+      source = match.group(1).lower()
+      signal = match.group(4).lower()
+
+      yield (path, (source, signal, time_type, geo_type, time_value))
+
+  @staticmethod
+  def is_header_valid(columns):
+    """Return whether the given pandas columns contains the required fields."""
+
+    return set(columns) >= CsvImporter.REQUIRED_COLUMNS
+
+  @staticmethod
+  def floaty_int(value):
+    """Cast a string to an int, even if it looks like a float.
+
+    For example, "-1" and "-1.0" should both result in -1. Non-integer floats
+    will cause `ValueError` to be reaised.
+    """
+
+    float_value = float(value)
+    int_value = round(float_value)
+    if float_value != int_value:
+      raise ValueError('not an int: "%s"' % str(value))
+    return int_value
+
+  @staticmethod
+  def maybe_apply(func, value):
+    """Apply the given function to the given value if not null-ish."""
+
+    if str(value).lower() not in ('', 'na', 'nan', 'inf', '-inf', 'none'):
+      return func(value)
+
+  @staticmethod
+  def extract_and_check_row(row, geo_type):
+    """Extract and return `RowValues` from a CSV row, with sanity checks.
+
+    Also returns the name of the field which failed sanity check, or None.
+
+    row: the pandas table row to extract
+    geo_type: the geographic resolution of the file
+    """
+
+    # use consistent capitalization (e.g. for states)
+    geo_id = row.geo_id.lower()
+
+    if geo_type in ('hrr', 'msa', 'dma'):
+      # these particular ids are prone to be written as ints -- and floats
+      geo_id = str(CsvImporter.floaty_int(geo_id))
+
+    # sanity check geo_id with respect to geo_type
+    if geo_type == 'county':
+      if len(geo_id) != 5 or not '01000' <= geo_id <= '80000':
+        return (None, 'geo_id')
+
+    elif geo_type == 'hrr':
+      if not 1 <= int(geo_id) <= 500:
+        return (None, 'geo_id')
+
+    elif geo_type == 'msa':
+      if len(geo_id) != 5 or not '10000' <= geo_id <= '99999':
+        return (None, 'geo_id')
+
+    elif geo_type == 'dma':
+      if not 450 <= int(geo_id) <= 950:
+        return (None, 'geo_id')
+
+    elif geo_type == 'state':
+      # note that geo_id is lowercase
+      if len(geo_id) != 2 or not 'aa' <= geo_id <= 'zz':
+        return (None, 'geo_id')
+
+    else:
+      return (None, 'geo_type')
+
+    # required float
+    value = float(row.val)
+
+    # optional nonnegative float
+    stderr = CsvImporter.maybe_apply(float, row.se)
+    if stderr is not None and stderr < 0:
+      return (None, 'se')
+
+    # optional not-too-small float
+    sample_size = CsvImporter.maybe_apply(float, row.sample_size)
+    if sample_size is not None and sample_size < 5:
```

Comment:
min sample size would be another i would recommend using a top-level constant for, and perhaps a comment as to why.

Reply:
done